baqend/chromeUXReports.ts

```diff
@@ -0,0 +1,349 @@
+      throw errors;
+    }
+
+    db.log.info(`Job ${job.id} completed.`);
+    const jobResults = await job.getQueryResults();
+    db.log.info('Job Result', jobResults);
+    return jobResults[0];
+  } catch (err) {
+    db.log.error('ERROR:', err);
+  }
+}
+
+/**
+ * Queries data from BigQuery
+ *
+ * @param db
+ * @param metric
+ * @param data
+ * @returns
+ */
+async function queryHistograms(db: baqend, metric: string, data: ChromeUXReportQueryData): Promise<any> {
+  const query =
+    "SELECT " +
+    "bin.start AS start, " +
+    "'all' AS device, " +
+    "SUM(bin.density) AS density " +
+    "FROM " +
+    "`chrome-ux-report.all." + data.year + data.month + "`, " +
+    "UNNEST(" + metric + ".histogram.bin) AS bin " +
+    "WHERE origin = '" + data.origin + "'" +
+    "GROUP BY start " +
+    "UNION ALL " +
+    "SELECT " +
+    "bin.start AS start, " +
+    "form_factor.name AS device, " +
+    "SUM(bin.density) AS density " +
+    "FROM " +
+    "`chrome-ux-report.all." + data.year + data.month + "`, " +
+    "UNNEST(" + metric + ".histogram.bin) AS bin " +
+    "WHERE origin = '" + data.origin + "' " +
+    "GROUP BY start, device " +
+    "ORDER BY start, device";
+
+  db.log.info(`Start Chrome User Experience Report for ${data.origin}`, {month: data.month, year: data.year});
+
+  return await asyncQuery(db, query)
+}
+
+/**
+ * Calculates the median for the given histogram
+ * @param histogram The given histogram
+ * @param totalDensity The total density reflects the share of a given form factor
+ * @returns
+ */
+function calculateMedian(histogram: model.ChromeUXReportData[], totalDensity: number): number {
+  let result: number = 0;
+
+  let cumulativeDistribution = 0;
+  for (const {start, density} of histogram) {
+    cumulativeDistribution = cumulativeDistribution + density;
+    if (cumulativeDistribution / totalDensity >= 0.5) {
+      result = start;
+      break;
+    }
+  }
+  return result;
+}
+
+async function getOrigin(db: baqend, data: QueriedParams): Promise<any> {
+  db.log.info('query origin', data.url);
+  const query =
+    "SELECT DISTINCT origin " +
+    "FROM `chrome-ux-report.all." + data.year + data.month + "` " +
+    "WHERE origin LIKE '%" + data.url + "'";
+
+  db.log.info(`Find origin in Chrome UX Reports for ${data.url}`, {month: data.month, year: data.year});
+  return asyncQuery(db, query);
+}
+
+/**
+ * Gets the origin from BigQuery to ensure available data
+ * @param db
+ * @param  data
+ * @returns origin or null, if queried url does not exist
+ */
+async function getOriginFromReports(db: baqend, data: QueriedParams): Promise<string | null> {
+  let url = new URL(data.url);
+  let host = url.host;
+  let addedWww = false;
+  // check for subdomain and if there is www
+  const match = host.match(/\./g);
+  if (match && match.length === 1 && !/^www/.test(host)) {
+    data.url = 'www.' + host;
+    addedWww = true;
+  }
+
+  let dataResult = await getOrigin(db, data);
```

Comment:
Hier lieber nur einen Query zu Google schicken. Konkret senden wir einfach den Host und benutzten dann REGEXP in dem SQL-Query an Google um uns die Ergebnisse zu holen. Wenn das Result größer als 1 sein sollte nehmen dir das, welches näher an der User-Eingabe ist (also entweder mit www oder ohne).